daemon/network.go

```diff
@@ -99,15 +100,51 @@ func (daemon *Daemon) getAllNetworks() []libnetwork.Network {
 	return daemon.netController.Networks()
 }
 
-func isIngressNetwork(name string) bool {
-	return name == "ingress"
+type ingressJob struct {
+	create *clustertypes.NetworkCreateRequest
+	ip     net.IP
 }
 
-var ingressChan = make(chan struct{}, 1)
+var (
+	ingressWorkerOnce  sync.Once
+	ingressJobsChannel chan *ingressJob
+	ingressWorkerStop  chan struct{}
+	ingressID          string
+	ingress            sync.Mutex
+)
+
+func (daemon *Daemon) startIngressWorker() {
+	ingress.Lock()
+	ingressWorkerStop = make(chan struct{})
+	ingressJobsChannel = make(chan *ingressJob, 100)
+	ingress.Unlock()
+	go func() {
+		for {
+			select {
+			case <-ingressWorkerStop:
+				ingressID = ""
+				return
+			case r := <-ingressJobsChannel:
+				if r.create != nil {
+					daemon.setupIngress(r.create, r.ip, ingressID)
+					ingressID = r.create.ID
+				} else {
+					daemon.releaseIngress(ingressID)
+					ingressID = ""
+				}
+			}
+		}
+	}()
+}
 
-func ingressWait() func() {
-	ingressChan <- struct{}{}
-	return func() { <-ingressChan }
+func (daemon *Daemon) stopIngressWorker() {
+	ingress.Lock()
+	if ingressWorkerStop != nil {
+		close(ingressWorkerStop)
+		ingressWorkerStop = nil
```

Comment:
This pattern may cause the goroutine in `startIngressWorker` to miss the channel close, because it might see the `nil` channel instead of the one that was closed. Also, the race detector will complain (and this might cause problems, depending on the internal implementation of `select`).